syntax = 'proto3';

package post_service;

option go_package = "./";

import "google/protobuf/timestamp.proto";


service PostService {
  rpc CreatePost (CreatePostRequest) returns (CreatePostResponse);
  rpc UpdatePost (UpdatePostRequest) returns (StatusResponse);
  rpc DeletePost (DeletePostRequest) returns (StatusResponse);
  rpc GetPostById (GetPostByIdRequest) returns (PostResponse);
  rpc ListPosts (ListPostsRequest) returns (ListPostsResponse);
}

message Post {
  int32 id = 1;
  int32 user_id = 2;
  string title = 3;
  string content = 4;
  google.protobuf.Timestamp created_at = 5;
  google.protobuf.Timestamp edited_at = 6;
}

message CreatePostRequest {
  int32 user_id = 1;
  string title = 2;
  string content = 3;
}

message CreatePostResponse {
    int32 id = 1;
}

message UpdatePostRequest {
  int32 id = 1;
  int32 user_id = 2;
  string title = 3;
  string content = 4;
}

message DeletePostRequest {
  int32 id = 1;
  int32 user_id = 2;
}

message GetPostByIdRequest {
  int32 id = 1;
}

message ListPostsRequest {
  int32 page = 1;
  int32 page_size = 2;
}

message ListPostsResponse {
  repeated Post posts = 1;
}

message PostResponse {
  Post post = 1;
  int32 status = 2;
}

message StatusResponse {
  int32 status = 1;
}

message EmptyResponse {}